### ISSUE TITLE
Embedded SSH key in node creation?

### ISSUE THREAD

**Issue by micolous** [exceedhl/toft#16]
https://github.com/exceedhl/toft/blob/master/lib/toft/node.rb#L9

There seems to be an SSH private key embedded in `toft` here.  It appears to be used to bootstrap some nodes.

Could this be used to take over nodes during or after the bootstrap process?

**Comment by exceedhl** [comment]
Yes, it is designed that way.

On Wed, Jan 8, 2014 at 3:05 PM, Michael Farrell notifications@github.comwrote:

> https://github.com/exceedhl/toft/blob/master/lib/toft/node.rb#L9
> 
> There seems to be an SSH private key embedded in toft here. It appears to
> be used to bootstrap some nodes.
> 
> Could this be used to take over nodes during or after the bootstrap
> process?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/exceedhl/toft/issues/16
> .

## 

Yours,
Liang

**Comment by micolous** [comment]
I can see that it is designed that way, it seems like a design flaw.

If a temporary key is required, generate one on-the-fly.  Or if there is an existing `ssh-agent`, populate the guest with that SSH key instead.

At the moment the key is left on the LXC container, so anyone who knows that this script is in use may gain root access on the LXC container -- the private key used is well-known.

This seems like a very bad security problem.

**Comment by exceedhl** [comment]
I agree with you if the the container is for production or any other
serious environment. However toft is designed to be a development and test
ancillary. Tools like docker seems a better choice if you want production
level container. Nonetheless it's good for toft to have some feature that
you described. Please feel free to modify the code and I am more than happy
to add those functionality.

On Fri, Jan 10, 2014 at 2:11 AM, Michael Farrell
notifications@github.comwrote:

> I can see that it is designed that way, it seems like a design flaw.
> 
> If a temporary key is required, generate one. Or if there is an existing
> ssh-agent, populate the guest with that SSH key instead.
> 
> At the moment the key is left on the LXC container, so anyone who knows
> that this script is in use may gain root access on the LXC container -- the
> private key used is well-known.
> 
> This seems like a very bad security problem.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/exceedhl/toft/issues/16#issuecomment-31959969
> .

## 

Yours,
Liang
